### PR TITLE
fix concurrent map write panics

### PR DESCRIPTION
**Pull request by infa-bsurber**
<!--
# General contribution criteria

Please have a look at our contribution guidelines: https://github.com/fluxcd/flux/blob/master/CONTRIBUTING.md
Particularly the sections about the:

 - DCO;
 - contribution workflow; and
 - how to get your fix accepted

To help the maintainers out when they're writing release notes, please
try to include a sentence or two here describing your change for end
users. See the CHANGELOG.md file in the top-level directory for examples.

Particularly for ground-breaking changes and new features, it's important to
make users and developers aware of what's changing and where those changes
were documented or discussed.

Even for smaller changes it's useful to see things documented as well, as it
gives everybody a chance to see at a glance what's coming up in the next
release. It makes the life of the project maintainer a lot easier as well.

The following short checklist can be used to make sure your PR is of good
quality, and can be merged easily:

- [x] if it resolves an issue;
      is a reference (i.e. #1) to this issue included?
- [N/A] if it introduces a new functionality or configuration flag;
      did you document this in the references or guides?
- [N/A] optional but much appreciated;
      do you think many users would profit from a dedicated setting
      for this functionality in the Helm chart?
-->

## Description

Solves issue #2917
- Unit test to prove that concurrent executions of getAllowedAndExistingNamespaces cause panics, which is leading to random panics in the wild. This test can be cleaned up after fix is tested if desired.
- Fix panics by using a concurrency-safe map or RWMutex 

## Unit Test Output Demonstrating Issue

```
go test -run TestConcurrentGetAllowedNamespaces

fatal error: concurrent map writes                                                                                                                 
                                                                                                                                               
goroutine 22418 [running]:                                                                                                                              
runtime.throw(0x2744a87, 0x15)                                                                            
        /usr/local/Cellar/go/1.12.3/libexec/src/runtime/panic.go:617 +0x72 fp=0xc03a1cdde0 sp=0xc03a1cddb0 pc=0x102fb22         
runtime.mapassign_faststr(0x24f1c80, 0xc0003ee960, 0x3af8ef4, 0x1, 0x0)                                      
        /usr/local/Cellar/go/1.12.3/libexec/src/runtime/map_faststr.go:291 +0x40f fp=0xc03a1cde48 sp=0xc03a1cdde0 pc=0x101536f
github.com/fluxcd/flux/pkg/cluster/kubernetes.(*Cluster).getAllowedAndExistingNamespaces(0xc000162500, 0x2b301e0, 0xc0000dc000, 0x0, 0x0, 0x0, 0x0, 0x0)
        /Users/bsurber/Go/src/github.com/infa-bsurber/flux/pkg/cluster/kubernetes/kubernetes.go:320 +0x669 fp=0xc03a1cdf78 sp=0xc03a1cde48 pc=0x235ab29
github.com/fluxcd/flux/pkg/cluster/kubernetes.TestConcurrentGetAllowedNamespaces.func1(0xc0000de300, 0xc000162500, 0xc0000de360)
        /Users/bsurber/Go/src/github.com/infa-bsurber/flux/pkg/cluster/kubernetes/kubernetes_test.go:74 +0x5b fp=0xc03a1cdfc8 sp=0xc03a1cdf78 pc=0x2379b2b
runtime.goexit()                                                       
        /usr/local/Cellar/go/1.12.3/libexec/src/runtime/asm_amd64.s:1337 +0x1 fp=0xc03a1cdfd0 sp=0xc03a1cdfc8 pc=0x105e4f1
created by github.com/fluxcd/flux/pkg/cluster/kubernetes.TestConcurrentGetAllowedNamespaces
        /Users/bsurber/Go/src/github.com/infa-bsurber/flux/pkg/cluster/kubernetes/kubernetes_test.go:72 +0x5bd

...
```

### Fix 1 performance (sync.Map)

```
go test -run TestConcurrentGetAllowedNamespaces 
PASS
ok      github.com/fluxcd/flux/pkg/cluster/kubernetes   19.955s
```

### Fix 2 performance (map of RWLocks)

```
still panicked actually, I guess maps can't be locked by key, and instead you have to lock the whole map when writing
```

### Fix 3 performance (single RWLock) <- my personal preferred option

```
go test -run TestConcurrentGetAllowedNamespaces
PASS
ok      github.com/fluxcd/flux/pkg/cluster/kubernetes   19.266s
```

### CODE REVIEW

File: pkg/cluster/kubernetes/kubernetes.go [hunk 1]
```diff
@@ -106,8 +106,9 @@ type Cluster struct {
 	syncErrors   map[resource.ID]error
 	muSyncErrors sync.RWMutex
 
-	allowedNamespaces map[string]struct{}
-	loggedAllowedNS   map[string]bool // to keep track of whether we've logged a problem with seeing an allowed namespace
+	allowedNamespaces   map[string]struct{}
+	loggedAllowedNS     map[string]bool // to keep track of whether we've logged a problem with seeing an allowed namespace
+	loggedAllowedNSLock *sync.RWMutex
```
Comment (hiddeco) [inline pkg/cluster/kubernetes/kubernetes.go]:
```suggestion
	loggedAllowedNSLock sync.RWMutex
```

With this you can remove `L131` :-)

**Comment by infa-bsurber** [comment]
made the fix per your comment and cleaned up the concurrency unit test after my final round of testing (don't want to add 20 seconds to every test run), then squashed the commits

**Comment by hiddeco** [comment]
Can you please rebase the branch on top of `master`?

**Comment by infa-bsurber** [comment]
oh wait, you said rebase, just a sec

**Comment by infa-bsurber** [comment]
there, rebased on upstream master

**Pull request merged**